### ISSUE TITLE
[tf.keras] Model.metrics_names bug in TensorFlow 2.2.0

### ISSUE THREAD

**Issue by briannemsick** [tensorflow/tensorflow#37990]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Y
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04):  Mac OS
- TensorFlow version (use command below): 2.2.0rc0
- Python version: 3.7

**Describe the current behavior**

Model.metrics_names returns an empty list (see below example) for a compiled model. This is new unexpected behavior as of TensorFlow 2.2.0 (not the case in TensorFlow 2.1.0)

These metrics names are important at compile time because they can be used to check against monitored quantities in callback. E.g. if a `ModelCheckpoint` callback is trying to monitor `val_lss` we can easily catch the typo before calling `model.fit` or finishing the first epoch of training.

**Describe the expected behavior**

Model.metric returns metrics and loss names (see below example).

**Standalone code to reproduce the issue** 
```python
import tensorflow as tf


inputs = tf.keras.layers.Input(shape=(3,))
outputs = tf.keras.layers.Dense(2, name="out_1")(inputs)
net = tf.keras.models.Model(inputs=inputs, outputs=outputs)
net.compile(optimizer="Adam", loss="mse", metrics=["mae"])

net.metrics_names = []  # TensorFlow 2.2.0rc2
net.metrics_names = ["loss", "mae"] # TensorFlow 2.1.0
```

**Comment by khimraj** [comment]
@briannemsick, In TF 2.2 , you need call `fit` or `train_on_batch` before you call  `net.metrics_names`.

**Comment by briannemsick** [comment]
That seems like a design regression. The eventual names are known at compile time.

Knowing the metrics names before called model.fit can prevent issues that would normally occur before the end of the epoch.

**Comment by kari-554** [comment]
Firstly we need to make proper layer in say sequential neural network as if before training the model if done the layering for doing the particular job we can easily use `metrics_names` as after the fit function or transform function can be used

**Comment by bjtho08** [comment]
This change also breaks several callbacks that rely on being able to read the `metrics `and `metrics_names` with the `on_train_begin()` hook. In my case, it breaks the [TensorFlow Addons](https://github.com/tensorflow/addons) `tqdm_progress_bar` [callback](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/tqdm_progress_bar.py). Any idea as to how I can implement a work-around to this issue. I had to upgrade to TF-2.2-r1 to fix a memory leak bug introduced in TF-2.1 and now I'm stuck with this instead.

**Comment by pavithrasv** [comment]
@briannemsick Thank you for pointing this out. This is definitely a breaking change but intended. Sorry about the lack of information on this on release notes and docs. We have fixed that now. 

We refactored compile/fit quite a bit and with that we made functional and subclassed model behaviors consistent. Because of this you will be able to use some of these model properties only after your model has been trained/evaluated on actual data.

**Comment by briannemsick** [comment]
In this case, a warning or exception when these attributes are accessed before train/eval feels warranted.

**Comment by bjtho08** [comment]
@pavithrasv Can I ask what is the reasoning behind this change? You mention refactoring to make things consistent, but how was the previous behavior inconsistent?

**Comment by pavithrasv** [comment]
@briannemsick We are looking into adding a warning as well. Thank you.

@bjtho08 Earlier you will be able to see metrics_name after compile call for functional model but not for subclassed model.

**Comment by pavithrasv** [comment]
@briannemsick there are constraints because of which we will not be able to add a warning but we have updated the docs with a note and examples to clarify the behavior.
